streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java

```diff
@@ -906,8 +906,8 @@ public void advanceWallClockTime(final Duration advance) {
         if (record == null) {
             throw new NoSuchElementException("Empty topic: " + topic);
         }
-        final K key = keyDeserializer.deserialize(record.topic(), record.key());
-        final V value = valueDeserializer.deserialize(record.topic(), record.value());
+        final K key = keyDeserializer.deserialize(record.topic(), record.headers(), record.key());
```

Comment:
Nice finding!

https://github.com/apache/kafka/blob/trunk/streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java#L805 has similar issue. Should we fix it as well?

Reply:
`readOutput` is deprecated. Thus not sure if it's worth to fix?

Reply:
> Thus not sure if it's worth to fix?

Personally, what we should add to ```readOutput``` is "deprecation" rather than "a bug". Hence, it is worthwhile to fix if the fix does not break anything.

Reply:
Not sure if I can follow? `readOutput` is marked as deprecated: https://github.com/apache/kafka/blob/trunk/streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java#L797

I can still fix it on the side, but nobody should use it any longer and thus the gain seems minimal.

Reply:
> but nobody should use it any longer and thus the gain seems minimal.

You are right. The benefit is too low.